mmselfsup/models/algorithms/interclr_moco.py

```diff
@@ -0,0 +1,426 @@
+                param_k.data.copy_(param_q.data)
+                param_k.requires_grad = False
+
+    @torch.no_grad()
+    def _momentum_update_key_encoder(self):
+        """Momentum update of the key encoder."""
+        for param_q, param_k in zip(self.encoder_q.parameters(),
+                                    self.encoder_k.parameters()):
+            param_k.data = param_k.data * self.momentum + \
+                           param_q.data * (1. - self.momentum)
+        if not self.share_neck:
+            for param_q, param_k in zip(self.inter_neck_q.parameters(),
+                                        self.inter_neck_k.parameters()):
+                param_k.data = param_k.data * self.momentum + \
+                               param_q.data * (1. - self.momentum)
+
+    @torch.no_grad()
+    def _dequeue_and_enqueue(self, keys):
+        """Update queue."""
+        # normalize
+        keys = nn.functional.normalize(keys, dim=1)
+        # gather keys before updating queue
+        keys = concat_all_gather(keys)
+
+        batch_size = keys.shape[0]
+
+        ptr = int(self.queue_ptr)
+        assert self.queue_len % batch_size == 0  # for simplicity
+
+        # replace the keys at ptr (dequeue and enqueue)
+        self.queue[:, ptr:ptr + batch_size] = keys.transpose(0, 1)
+        ptr = (ptr + batch_size) % self.queue_len  # move pointer
+
+        self.queue_ptr[0] = ptr
+
+    def contrast_intra(self, q, k):
+        """Intra-image invariance learning."""
+        q = nn.functional.normalize(q, dim=1)
+        k = nn.functional.normalize(k, dim=1)
+        # compute logits
+        # Einstein sum is more intuitive
+        # positive logits: Nx1
+        pos_logits = torch.einsum('nc,nc->n', [q, k]).unsqueeze(-1)
+        # negative logits: NxK
+        neg_logits = torch.einsum('nc,ck->nk',
+                                  [q, self.queue.clone().detach()])
+
+        if self.intra_cos:
+            cosine = pos_logits.clone()
+            phi = cosine - self.intra_cos_margin
+            pos_logits.copy_(phi)
+        if self.intra_arc:
+            cosine = pos_logits.clone()
+            sine = torch.sqrt((1.0 - cosine**2).clamp(0, 1))
+            phi = cosine * math.cos(self.intra_arc_margin) - sine * math.sin(
+                self.intra_arc_margin)
+            if self.intra_arc_margin < 0:
+                phi = torch.where(
+                    cosine < math.cos(self.intra_arc_margin), phi, cosine +
+                    math.sin(self.intra_arc_margin) * self.intra_arc_margin)
+            else:
+                phi = torch.where(
+                    cosine > math.cos(math.pi - self.intra_arc_margin), phi,
+                    cosine - math.sin(math.pi - self.intra_arc_margin) *
+                    self.intra_arc_margin)
+            pos_logits.copy_(phi)
+
+        losses = self.head(pos_logits, neg_logits)
+
+        return losses
+
+    def contrast_inter(self, q, idx):
```

Comment:
the same question as function `contrast_intra`, besides better to add more docstrings to explain these two functions

Reply:
same as `contrast_intra`